
#### Purpose ####
# This script was used to create CpG summary tables extracted from the 'Cyto_Summary' reports
# generated by bismark. The data was run through a standard methylKit protocol to create single data
# matrices (CpGs x # of Samples), bed files, and perform differential methylation using the methylKit approach.

#### Libraries ####
 # Downloaded from Bioconductor using BiocManager::install()
library(methylKit) # DNA methylation manipulation
library(genomation)
library(data.table)# fwrite for faster data writing
library(matrixStats) # rowMedians function
library(readr) 
library(dplyr) 
library(ggplot2)

#### Environment ####
sessionInfo()
# R version 3.6.0 (2019-04-26)
# Platform: x86_64-redhat-linux-gnu (64-bit)
# Running under: CentOS Linux 7 (Core)
# 
# Matrix products: default
# BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so
# 
# locale:
#   [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
# [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
# [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
# [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
# [9] LC_ADDRESS=C               LC_TELEPHONE=C            
# [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
# 
# attached base packages:
#   [1] grid      parallel  stats4    stats     graphics  grDevices utils    
# [8] datasets  methods   base     
# 
# other attached packages:
#   [1] genomation_1.16.0    methylKit_1.10.0     GenomicRanges_1.36.1
# [4] GenomeInfoDb_1.20.0  IRanges_2.18.3       S4Vectors_0.22.1    
# [7] BiocGenerics_0.30.0 
# 
# loaded via a namespace (and not attached):
#   [1] Biobase_2.44.0              splines_3.6.0              
# [3] R.utils_2.9.2               gtools_3.8.1               
# [5] assertthat_0.2.1            BSgenome_1.52.0            
# [7] GenomeInfoDbData_1.2.1      Rsamtools_2.0.3            
# [9] impute_1.58.0               numDeriv_2016.8-1.1        
# [11] pillar_1.4.3                backports_1.1.5            
# [13] lattice_0.20-38             glue_1.3.1                 
# [15] limma_3.40.6                bbmle_1.0.22               
# [17] XVector_0.24.0              qvalue_2.2.0               
# [19] colorspace_1.4-1            Matrix_1.2-17              
# [21] R.oo_1.23.0                 plyr_1.8.5                 
# [23] XML_3.99-0.3                pkgconfig_2.0.3            
# [25] emdbook_1.3.11              zlibbioc_1.30.0            
# [27] purrr_0.3.3                 mvtnorm_1.0-12             
# [29] scales_1.1.0                BiocParallel_1.18.1        
# [31] tibble_2.1.3                mgcv_1.8-28                
# [33] ggplot2_3.2.1               seqPattern_1.16.0          
# [35] SummarizedExperiment_1.14.1 lazyeval_0.2.2             
# [37] magrittr_1.5                crayon_1.3.4               
# [39] mclust_5.4.5                R.methodsS3_1.7.1          
# [41] nlme_3.1-139                MASS_7.3-51.4              
# [43] tools_3.6.0                 data.table_1.12.8          
# [45] hms_0.5.3                   lifecycle_0.1.0            
# [47] matrixStats_0.55.0          gridBase_0.4-7             
# [49] stringr_1.4.0               munsell_0.5.0              
# [51] plotrix_3.7-7               DelayedArray_0.10.0          
# [53] Biostrings_2.52.0           compiler_3.6.0             
# [55] fastseg_1.30.0              rlang_0.4.2                
# [57] RCurl_1.98-1.1              bitops_1.0-6               
# [59] gtable_0.1.2                reshape2_1.4.3             
# [61] R6_2.4.1                    GenomicAlignments_1.20.1   
# [63] dplyr_0.8.3                 rtracklayer_1.44.4         
# [65] bdsmatrix_1.3-4             zeallot_0.1.0              
# [67] KernSmooth_2.23-15          readr_1.3.1                
# [69] stringi_1.4.5               Rcpp_1.0.3                 
# [71] vctrs_0.2.1                 tidyselect_0.2.5           
# [73] coda_0.19-3     

    #### Functions ####

## Basic function for creating a sub directory and also checking if it exists. 
newDir <- function(mainDir,subDir){
  if (file.exists(subDir)){
    setwd(file.path(mainDir, subDir))
  } else {
    dir.create(file.path(mainDir, subDir))
    setwd(file.path(mainDir, subDir))
  }
}

## Function which combins the calculateDiffMeth() and getMethylDiff() functions and creates writes the table outputs
 # It currently defaults to calculate overdispersion using the "MN" method and using a "Chisq" test, and 15 compute cores.
 # These could easily be changed in the function or added as arguments.
 # It does allow the user to specify the threshold of significant DMLs by providing the diff.methyl cut off and qvalue.
 # x : methylKit object made with the union function
 # dM :  difference in methylation among comparison.
 # q : q.value threshold.
 # cov : vector or matrix of covariates (this is used when all the data is examined, i.e. timepoint is included as covariate when looking
 # as temperature).
 # type : "DML" or "DMR" used to describe the state of the methylation datae (individual loci or regional)
 # comparison : "all","tp09","tp80" used to describe the specific comparison being performed
 # output :  output directory for the files.

performDiffMethSteps <- function(x,dM=50,q=0.01,cov=NULL,type="DML",comparison="all",output=baseOutputDir){
  newDir(output,paste0("dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2]))

  ### ChiSq Approach with overdispersion correction
  app <- "ChiSq_overDispCorrected"
  y<-calculateDiffMeth(x,
                       covariates=cov,
                       overdispersion="MN",test="Chisq",mc.cores=15)
  saveRDS(y,paste0(type,"_",comparison,app,".RData"))
  fwrite(y,paste0(type,"_",comparison,app,".csv"),sep = ",")
  
  # Multicomparison correction and critical % Diff. Threshold
  y_getDiff <- getMethylDiff(y,difference=50,qvalue=0.01)
  saveRDS(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.RData"))
  fwrite(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.csv"),sep = ",")
  # Hyper and Hypomethylation
  yhyper <- y_getDiff[y_getDiff$meth.diff > 0,]
  saveRDS(yhyper,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hyper.RData"))
  yhypo<- y_getDiff[y_getDiff$meth.diff < 0,]
  saveRDS(yhypo,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hypo.RData"))
  
  ### ChiSq Approach without overdispersion 
  app <- "ChiSq_uncorrected"
  y<-calculateDiffMeth(x,
                       covariates=cov,test="Chisq",mc.cores=15)
  saveRDS(y,paste0(type,"_",comparison,app,".RData"))
  fwrite(y,paste0(type,"_",comparison,app,".csv"),sep = ",")
  # Multicomparison correction and critical % Diff. Threshold
  y_getDiff <- getMethylDiff(y,difference=50,qvalue=0.01)
  saveRDS(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.RData"))
  fwrite(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.csv"),sep = ",")
  # Hyper and Hypomethylation
  yhyper <- y_getDiff[y_getDiff$meth.diff > 0,]
  saveRDS(yhyper,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hyper.RData"))
  yhypo<- y_getDiff[y_getDiff$meth.diff < 0,]
  saveRDS(yhypo,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hypo.RData"))
  
  ### DSS method
  # app <- "DSS"
  # y<-calculateDiffMethDSS(meth,mc.cores=15)
  # saveRDS(y,paste0(type,"_",comparison,app,".RData"))
  # fwrite(y,paste0(type,"_",comparison,app,".csv"),sep = ",")
  # # Multicomparison correction and critical % Diff. Threshold
  # y_getDiff <- getMethylDiff(y2,difference=50,qvalue=0.01)
  # saveRDS(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.RData"))
  # fwrite(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.csv"),sep = ",")
  # # Hyper and Hypomethylation
  # yhyper <- y_getDiff[y_getDiff$meth.diff > 0,]
  # saveRDS(yhyper,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hyper.RData"))
  # yhypo<- y_getDiff[y_getDiff$meth.diff < 0,]
  # saveRDS(yhypo,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hypo.RData"))
}

## Function for creates a simple bed file with chr,start,end,label from the cpgs matrix files generated from methylkit.
 # It defaults to putting this file in a folder called bed in a output directory specified by the user.
 # `x` : matrix with cpgs by row and labels columns including chr,start,end,and label (this can be predefined as anything in the matrix).
 # `start_offset` and `end_offset` can manually change the range of base offsets needed given the start and end base (this is to adjust
 # for the fact that methylkit and bismark defines the cpg as a single based even after destranding). The default behavior will 
 # adjust for this and create the appropriate start and end for a CpG.
 # `type` :  file name
 # `output` :  output directory

bedFileCreate <- function(x,start_offset=-1,end_offset=1,type = "test",output=NULL){
  newDir(output,"bed")
  y <- mutate(x,label=paste0(chr,"_",start,"_",end),
              start = start+start_offset, end = end+end_offset) %>% 
    select(chr, start, end,label) %>% 
    mutate_if(is.numeric, as.integer) #Save as a BED file, and avoid writing information in scientific notation
  fwrite(y,paste0(type,".bed"),
         sep = '\t',
         col.names = FALSE) #Save data as a BED file
}

## Methylation plots by treatment
# `x` : methylObject generated by methylKit unite.
# `fileName` : Name on the file to represent the comparison for the specific set of DMLs
# `label` :  dataframe that contains a column of sample IDs and column of treatments.

methylationPlotByTreatment <- function(x,fileName,label){
  temp_t = matrix(unlist(getData(x)[,seq(5,ncol(x),by=3)]),ncol=nrow(label))
  temp_c = matrix(unlist(getData(x)[,seq(6,ncol(x),by=3)]),ncol=nrow(label))
  temp_b = temp_c/temp_t
  print("Making figures...")
  for(i in 1:nrow(x)){
    plot_df <- data.frame(ID=label$ID,meta=label$var,b=temp_b[i,],cov=temp_t[i,]) 
    print(paste0("Plot ",i," of ",nrow(x)))
    # Basic Boxplot by treatment
    p1 <- ggplot(plot_df,aes(x=meta,y=b,group=meta,col=meta))+ geom_boxplot() + labs(y="Prop. Methylation")
    # Methylation for each individual
    p2 <- ggplot(plot_df,aes(x=ID,y=b,group=meta,col=meta))+ geom_point() + labs(y="Prop. Methylation",x="Individuals") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1))
    # Coverage plot
    p3 <- ggplot(plot_df,aes(x=cov,y=b,col=meta))+ geom_point() + 
      labs(y="Prop. Methylation",x="Coverage")
    # Coverage histogram
    p4 <- ggplot(plot_df,aes(x=cov,fill=meta)) + geom_histogram(bins=30)
    final<-plot_grid(p1,p2,p3,p4)#,nrow=2 # p1,p2,
    save_plot(paste0("DMLbyTrt_",fileName,"_",getData(x)[i,1],"_",getData(x)[i,2],"_beta.png"),final, ncol = 2)
  }
}

#### Data ####

## Set working directory
#setwd("/shared_lab/20180226_RNAseq_2017OAExp/DNAm/processed_samples/03_CytoSummaries")

# Paths to input and output directories
inputDir_meta <- "~/Github/AE17_Cvirginica_MolecularResponse"
inputDir_CytoSummaries <- "/path/to/CytoReportFiles"
# Outputs on local directory
outputDir <- "/home/downeyam/Github/AE17_Cvirginica_MolecularResponse/data/MBDBS_seq"
setwd(inputDir_meta)
meta <- readRDS("data/meta/metadata_20190811.RData")

# Temp for cluster
inputDir_meta <- "/shared_lab/20180226_RNAseq_2017OAExp/DNAm/"
inputDir_CytoSummaries <- "/shared_lab/20180226_RNAseq_2017OAExp/DNAm/processed_samples/03_CytoSummaries"
setwd(inputDir_meta)
meta <- readRDS("metadata/metadata_20190811.RData")
setwd(inputDir_CytoSummaries)

# Meta data
meta_final <- meta[meta$ID != "17099",]
# Convert treatment into binary (0,1)
trt <- ifelse(meta_final$treatment==400,0,1)
timeVec <- ifelse(meta_final$Time=="09",0,1)

# Cytosine Summary Reports from Bismark
setwd(inputDir_CytoSummaries)
file.list <- list.files(pattern="17")

# Removing sample 17099 which sequenced poorly
file.list <- file.list[-11]
ID <- meta_final$ID

#### Read in files ####

## Stores all CpGs (mincov -  minimum coverage = 0)
myobj <- methRead(as.list(file.list),
                 sample.id=as.list(ID),
                 assembly="t2",
                 treatment=trt,mincov = 0,
                 pipeline='bismarkCytosineReport')
# Here I want to be able to see all CpGs, and we also filter in the next step

# Here I change my working directory since we will be saving our outputs from here on out.

newDir(inputDir_CytoSummaries,"methylObjFiles")
newDir(inputDir_CytoSummaries,"temp")
baseOutputDir <- getwd()
#### Filter samples ####
# Defaults used
# Needs at least 5x coverage
# And remove loci with coverage > 200 (potential PCR bias)
filtered.myobj <- filterByCoverage(myobj,lo.count=5,lo.perc=NULL,
                                hi.count=200,hi.perc=NULL)
#### Normalize counts ####
 # The function normalizes coverage values between samples using a scaling factor derived 
 # from differences between mean or median of coverage distributions
 ## Default is median
norm.myobj <- normalizeCoverage(filtered.myobj,method="median") 
saveRDS(norm.myobj,"methylKitObj_all_medianNormalizedUnfiltered.RData")
fwrite(norm.myobj,"methylKitObj_all_medianNormalizedUnfiltered.csv",sep = ",")

#### Unite will combine samples ####
setwd(baseOutputDir)
 # Function will shrink dataframe of each individual to only 
 # those shared by all individuals (i.e. only loci with 5x for all samples)
  
## Saving all CpGs (no filter)
meth_all <- unite(myobj, destrand=TRUE)
#mKitPlot(meth_all,"allTimepoints_noFilter")
saveRDS(meth_all,"methylKitObj_all_allCpGs_united.RData")
fwrite(meth_all,"methylKitObj_all_allCpGs_united.csv",sep = ",")

## All time points (5x coverage)
meth <- unite(norm.myobj, destrand=TRUE)
# Visualization
#mKitPlot(meth,"allTimepoints")
saveRDS(meth,"methylKitObj_all_cov5Filtered_united.RData")
fwrite(meth,"methylKitObj_all_cov5Filtered_united.csv",sep = ",")

# Read in data (optional)
# newDir(inputDir_CytoSummaries,"methylObjFiles")
# baseOutputDir <- getwd()
# meth_all_tile <- readRDS("methylKitObj_all_cov5Filtered_united.RData")

## All time points with time as variable 
meth_time <- meth
meth_time@treatment <-  timeVec

#### Use reorganize to subset data by timepoints ####

## Day 9 (5x coverage)
meta_9 <- meta_final[meta_final$Time == "09",]
trt9 <-  ifelse(meta_9$treatment==400,0,1)
meth9 <- reorganize(meth,sample.ids=meta_9$ID,treatment=trt9)
# Visualization
#mKitPlot(meth9,"timepoint9")
saveRDS(meth9,"methylKitObj_tp9_cov5Filtered_united.RData")
fwrite(meth9,"methylKitObj_tp9_cov5Filtered_united.csv",sep = ",")

## Day 80 (5x coverage)
meta_80 <- meta_final[meta_final$Time == "80",]
trt80 <- ifelse(meta_80$treatment==400,0,1)
meth80 <- reorganize(meth,sample.ids=meta_80$ID,treatment=trt80)
# Visualization
#mKitPlot(meth80,"timepoint80")
saveRDS(meth80,"methylKitObj_tp80_cov5Filtered_united.RData")
fwrite(meth80,"methylKitObj_tp80_cov5Filtered_united.csv",sep = ",")

## Control Treatment - Day 9 vs Day 80 (5x coverage)
meta_C <- meta_final[meta_final$treatment == 400,]
trtC <- ifelse(meta_C$Time=="09",0,1)
methC <- reorganize(meth_time,sample.ids=meta_C$ID,treatment=trtC)
# Visualization
saveRDS(methC,"methylKitObj_C_cov5Filtered_united.RData")
fwrite(methC,"methylKitObj_C_cov5Filtered_united.csv",sep = ",")

## Exposed Treatment - Day 9 vs Day 80 (5x coverage)
meta_E <- meta_final[meta_final$treatment == 2800,]
trtE <- ifelse(meta_E$Time=="09",0,1)
methE <- reorganize(meth_time,sample.ids=meta_E$ID,treatment=trtE)
# Visualization
saveRDS(methE,"methylKitObj_E_cov5Filtered_united.RData")
fwrite(methE,"methylKitObj_E_cov5Filtered_united.csv",sep = ",")

#### Create separate counts matrices and beta (methylC/totalC) ####
newDir(baseOutputDir,"count_matrix_5X")

meth_all_counts <- as.matrix(getData(meth)[,5:73])
#Generate the row names 
rName <- paste0(getData(meth)[,1],"_",getData(meth)[,2])
# Generate the counts matrix
meth_total <-meth_all_counts[,seq(from=1,to=69,by=3)]
meth_M <-meth_all_counts[,seq(from=2,to=69,by=3)]
meth_uM <-meth_all_counts[,seq(from=3,to=69,by=3)]
meth_beta <- meth_M/meth_total

rownames(meth_total) <- rName
rownames(meth_M) <- rName
rownames(meth_uM) <- rName
rownames(meth_beta) <- rName

fwrite(meth_total,"methylKitObj_all_cov5Filtered_united_totalCounts.csv",sep = ",")
fwrite(meth_M,"methylKitObj_all_cov5Filtered_united_MethylCCounts.csv",sep = ",")
fwrite(meth_uM,"methylKitObj_all_cov5Filtered_united_unMethylCCounts.csv",sep = ",")
fwrite(meth_beta,"methylKitObj_all_cov5Filtered_united_beta.csv",sep = ",")

## Print out histogram of beta value distribution for all CpGs with at least 5x
rowMdn<-rowMedians(meth_beta)
png("Hist_Beta_5x_methylKit.png")
hist(rowMdn)
dev.off()

#### Create a summary of methylation for each CpG ####
# Use the beta matrix created in previous section
b <- as.matrix(meth_beta)
# Mean methylation all samples
mean_beta <- rowMeans(b)
sd_beta <- rowSds(b)
cv_beta <- sd_beta/mean_beta

# Individual Treatment x Time level combination summaries
# Mean methylation Control day 9
mean_9C <- rowMeans(b[,meta_final$SFV == "09.400"]) # mean
sd_9C <- rowSds(b[,meta_final$SFV == "09.400"]) # standard deviation
cv_9C <- sd_9C/mean_9C # coefficient of variation
# Mean methylation Exposed day 9
mean_9E <- rowMeans(b[,meta_final$SFV == "09.2800"])
sd_9E <- rowSds(b[,meta_final$SFV == "09.2800"])
cv_9E <- sd_9E/mean_9E
# Mean methylatoin Control day 80
mean_80C <- rowMeans(b[,meta_final$SFV == "80.400"])
sd_80C <- rowSds(b[,meta_final$SFV == "80.400"])
cv_80C <- sd_80C/mean_80C
# Mean methylation Exposed day 80
mean_80E <- rowMeans(b[,meta_final$SFV == "80.2800"])
sd_80E <- rowSds(b[,meta_final$SFV == "80.400"])
cv_80E <- sd_80E/mean_80E

# Coefficient of Variation among treatments (all levels)
mean_AmongTrt <- cbind(mean_9C,mean_9E,mean_80C,mean_80E)
sd_mean_AmongTrt <- rowSds(mean_AmongTrt)
mean_mean_AmongTrt <- rowMeans(mean_AmongTrt)
cv_mean_AmongTrt <- sd_mean_AmongTrt/mean_mean_AmongTrt

# Difference in Methylation trt
mean_C <- rowMeans(b[,meta_final$Treatment ==400])
mean_E <- rowMeans(b[,meta_final$Treatment == 2800])
diff_Trt <- mean_E-mean_C

# Difference in Methylation time
mean_9 <- rowMeans(b[,meta_final$Time == "09"])
mean_80 <- rowMeans(b[,meta_final$Time == "80"])
diff_Time <- mean_E-mean_C

# Difference in Methylation tp9
diff_tp9_Trt <- mean_9E-mean_9C
# Difference in Methylation tp80
diff_tp80_Trt <- mean_80E-mean_80C
# Creat matrix of all CpG summary stats
meth_summary <- cbind(mean_beta,
                      mean_9C,mean_9E,mean_80C,mean_80E,
                      sd_9C,sd_9E,sd_80C,sd_80E,sd_mean_AmongTrt,
                      cv_9C,cv_9E,cv_80C,cv_80E,cv_mean_AmongTrt,
                      diff_Trt,diff_Time,diff_tp9_Trt,diff_tp80_Trt)
meth_summary[is.na(meth_summary)]<-0 
# in cases where the mean of treatmentxtime combination is 0 for the CV calculation will produce an NA,
# we want to change these to 0s.

#### Read in files from saved folder (optional)####
# If you want to start with the united methyKit Objects 
 setwd(outputDir)
 baseOutputDir <- getwd()
 meth_all <- readRDS("methylKitObj_all_cov5Filtered_united.RData")
 meth <- readRDS("methylKitObj_all_cov5Filtered_united.RData")
 meth9 <- readRDS("methylKitObj_tp9_cov5Filtered_united.RData")
 meth80 <- readRDS("methylKitObj_tp80_cov5Filtered_united.RData")

#### Differential Methyl ####
# Differential Methylation done with the calculateDiffMeth() function, which takes a logisitic regression approach
# Critical criteria for significant
  # q value : 0.01
  # Diff methylation : >= 50%

cov = data.frame(tp=meta_final$timepoint)
# All trt
performDiffMethSteps(meth,cov=cov,type="DML",output=baseOutputDir)
# All tp 9
performDiffMethSteps(meth9,type="DML",comparison="tp9",output=baseOutputDir)
# All tp80
performDiffMethSteps(meth80,type="DML",comparison="tp80",output=baseOutputDir)
# All timepoint
cov = data.frame(tp=meta_final$treatment)
performDiffMethSteps(meth_time,cov=cov,type="DML",comparison="all_timepoint",output=baseOutputDir)
# C
performDiffMethSteps(methC,type="DML",comparison="C",output=baseOutputDir)
# E
performDiffMethSteps(methE,type="DML",comparison="E",output=baseOutputDir)

#### Create 'bed' file formats for these outputs ####
newDir(baseOutputDir,"bed")
## All CpGs (not necessarily covered)
meth_all <- readRDS(paste0(baseOutputDir,"/methylKitObj_all_allCpGs_united.RData"))
bedFileCreate(x=meth_all,type="CpG_allCpGs",output=baseOutputDir)
## All CpGs with 5x coverage with summary 
meth <- readRDS(paste0(baseOutputDir,"/methylKitObj_all_cov5Filtered_united.RData"))
CpG_5xCov_bed <- mutate(meth,label=paste0(chr,"_",start),start = start-1, end = end + 1) %>% 
  select(chr, start, end,label) %>% 
  mutate_if(is.numeric, as.integer) #Save as a BED file, and avoid writing information in scientific notation
CpG_5xCov_bed <- data.frame(CpG_5xCov_bed,meth_summary)
fwrite(CpG_5xCov_bed,
            "CpG_5xCov.bed",
            sep = '\t',
            col.names = FALSE) #Save data as a BED file
## All Regions 
meth_all_tile <- readRDS(paste0(baseOutputDir,"/methylKitObj_all_medianNormalizedUnfiltered_united_tiled.RData"))
bedFileCreate(x=meth_all_tile,type="Region_100BpTile",output=baseOutputDir)

## DMLs all timepoints
files <- list.files(paste0(baseOutputDir,"/dM50_q01"))
subsetSample <- c("all_d","tp9","tp09","tp80","all_t","C","E")
for(i in 1:length(subsetSample)){
  target <- files[grep(subsetSample[i],files)]
  target <- target[grep("DML",target)]
  temp <- target[grep(".RData",target)]
  if(length(temp)>0){
    for(j in 1:length(temp)){
      x <- readRDS(paste0(baseOutputDir,"/dM50_q01/",temp[j]))
      name <- strsplit(temp[j],"\\.")[[1]][1]
      bedFileCreate(x=x,type = name,output=baseOutputDir)
    }
  }
  target <- files[grep(subsetSample[i],files)]
  target <- target[grep("DMR",target)]
  temp <- target[grep(".RData",target)]
  if(length(temp)>0){
    for(j in 1:length(temp)){
      x <- readRDS(paste0(baseOutputDir,"/dM50_q01/",temp[j]))
      name <- strsplit(temp[j],"\\.")[[1]][1]
      bedFileCreate(x=x,type = name,output=baseOutputDir)
    }
  }
}

#### DMRs - Tiling (optional) #####
# This will section will perform a similar differential methylation analysis, 
# but summarizing CpGs into 100bp windows.
??tileMethylCounts()
tiles <- tileMethylCounts(norm.myobj,win.size=100,step.size=100)
saveRDS(tiles,"methylKitObj_all_medianNormalizedUnfiltered_tiled.RData")

## Unite samples
# All points
meth_all_tile <- unite(tiles, destrand=TRUE)
saveRDS(meth_all_tile,"methylKitObj_all_medianNormalizedUnfiltered_united_tiled.RData")

# Read in data (optional)
# newDir(inputDir_CytoSummaries,"methylObjFiles")
# baseOutputDir <- getwd()
# meth_all_tile <- readRDS("methylKitObj_all_medianNormalizedUnfiltered_united_tiled.RData")

# All points with time rather than OA as treatment
meth_all_time_tile <- meth_all_tile
meth_all_time_tile@treatment <- timeVec 
# TP 9
meta_9 <- meta_final[meta_final$Time == "09",]
trt9 <-  ifelse(meta_9$treatment== 400,0,1)
meth9_tile <- reorganize(meth_all_tile,sample.ids=meta_9$ID,treatment=trt9)
# TP 80
meta_80 <- meta_final[meta_final$Time == "80",]
trt80 <- ifelse(meta_80$treatment==400,0,1)
meth80_tile <- reorganize(meth_all_tile,sample.ids=meta_80$ID,treatment=trt80)
# Control Treatment (day 9  vs day 80)
meta_C <- meta_final[meta_final$treatment == 400,]
trtC <- ifelse(meta_C$Time=="09",0,1)
methC_tile <- reorganize(meth_all_time_tile,sample.ids=meta_C$ID,treatment=trtC)
# Exposure Treatment (day 9  vs day 80)
meta_E <- meta_final[meta_final$treatment == 2800,]
trtE <- ifelse(meta_E$Time=="09",0,1)
methE_tile <- reorganize(meth_all_time_tile,sample.ids=meta_E$ID,treatment=trtE)

#### Differential Methylation Tiles (optional) ####
# Get all loci that are differentially methylated
# with a difference of at least 50% and a qvluae=0.01

cov = data.frame(tp=meta_final$timepoint)
# All trt
performDiffMethSteps(meth_all_tile,cov=cov,type="DMR",output=baseOutputDir)
# All tp 9
performDiffMethSteps(meth9_tile,type="DMR",comparison="tp09",output=baseOutputDir)
# All tp80
performDiffMethSteps(meth80_tile,type="DMR",comparison="tp80",output=baseOutputDir)
# All timepoint
cov = data.frame(tp=meta_final$treatment)
performDiffMethSteps(meth_all_time_tile,cov=cov,type="DMR",comparison="all_timepoint",output=baseOutputDir)
# C
performDiffMethSteps(methC_tile,type="DMR",comparison="C",output=baseOutputDir)
# E
performDiffMethSteps(methE_tile,type="DMR",comparison="E",output=baseOutputDir)


