
#### Purpose ####
# This script was used to create CpG summary tables extracted from the 'Cyto_Summary' reports
# generated by bismark. The data was run through a standard methylKit protocol to create single data
# matrices (CpGs x # of Samples), bed files, and perform differential methylation using the methylKit approach.

#### Libraries ####
 # Downloaded from Bioconductor using BiocManager::install()
library(methylKit) # DNA methylation manipulation
library(genomation)
library(data.table)# fwrite for faster data writing
library(matrixStats) # rowMedians function
library(readr) 
library(dplyr) 
library(ggplot2)
library(cowplot)
library(corrplot)
library(plyr)
library(grid)
library(gridExtra)
library(RColorBrewer)
pal <- brewer.pal(n = 12, name = 'Paired')
col_perm <- c(pal[1],pal[5],pal[2],pal[6])

#### Environment ####
sessionInfo()
# R version 3.6.0 (2019-04-26)
# Platform: x86_64-redhat-linux-gnu (64-bit)
# Running under: CentOS Linux 7 (Core)
# 
# Matrix products: default
# BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so
# 
# locale:
#   [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
# [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
# [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
# [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
# [9] LC_ADDRESS=C               LC_TELEPHONE=C            
# [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
# 
# attached base packages:
#   [1] grid      parallel  stats4    stats     graphics  grDevices utils    
# [8] datasets  methods   base     
# 
# other attached packages:
#   [1] genomation_1.16.0    methylKit_1.10.0     GenomicRanges_1.36.1
# [4] GenomeInfoDb_1.20.0  IRanges_2.18.3       S4Vectors_0.22.1    
# [7] BiocGenerics_0.30.0 
# 
# loaded via a namespace (and not attached):
#   [1] Biobase_2.44.0              splines_3.6.0              
# [3] R.utils_2.9.2               gtools_3.8.1               
# [5] assertthat_0.2.1            BSgenome_1.52.0            
# [7] GenomeInfoDbData_1.2.1      Rsamtools_2.0.3            
# [9] impute_1.58.0               numDeriv_2016.8-1.1        
# [11] pillar_1.4.3                backports_1.1.5            
# [13] lattice_0.20-38             glue_1.3.1                 
# [15] limma_3.40.6                bbmle_1.0.22               
# [17] XVector_0.24.0              qvalue_2.2.0               
# [19] colorspace_1.4-1            Matrix_1.2-17              
# [21] R.oo_1.23.0                 plyr_1.8.5                 
# [23] XML_3.99-0.3                pkgconfig_2.0.3            
# [25] emdbook_1.3.11              zlibbioc_1.30.0            
# [27] purrr_0.3.3                 mvtnorm_1.0-12             
# [29] scales_1.1.0                BiocParallel_1.18.1        
# [31] tibble_2.1.3                mgcv_1.8-28                
# [33] ggplot2_3.2.1               seqPattern_1.16.0          
# [35] SummarizedExperiment_1.14.1 lazyeval_0.2.2             
# [37] magrittr_1.5                crayon_1.3.4               
# [39] mclust_5.4.5                R.methodsS3_1.7.1          
# [41] nlme_3.1-139                MASS_7.3-51.4              
# [43] tools_3.6.0                 data.table_1.12.8          
# [45] hms_0.5.3                   lifecycle_0.1.0            
# [47] matrixStats_0.55.0          gridBase_0.4-7             
# [49] stringr_1.4.0               munsell_0.5.0              
# [51] plotrix_3.7-7               DelayedArray_0.10.0          
# [53] Biostrings_2.52.0           compiler_3.6.0             
# [55] fastseg_1.30.0              rlang_0.4.2                
# [57] RCurl_1.98-1.1              bitops_1.0-6               
# [59] gtable_0.1.2                reshape2_1.4.3             
# [61] R6_2.4.1                    GenomicAlignments_1.20.1   
# [63] dplyr_0.8.3                 rtracklayer_1.44.4         
# [65] bdsmatrix_1.3-4             zeallot_0.1.0              
# [67] KernSmooth_2.23-15          readr_1.3.1                
# [69] stringi_1.4.5               Rcpp_1.0.3                 
# [71] vctrs_0.2.1                 tidyselect_0.2.5           
# [73] coda_0.19-3     

#### Functions ####

## Basic function for creating a sub directory and also checking if it exists. 
newDir <- function(mainDir,subDir){
  if (file.exists(subDir)){
    setwd(file.path(mainDir, subDir))
  } else {
    dir.create(file.path(mainDir, subDir))
    setwd(file.path(mainDir, subDir))
  }
}

## Function which combins the calculateDiffMeth() and getMethylDiff() functions and creates writes the table outputs
 # It currently defaults to calculate overdispersion using the "MN" method and using a "Chisq" test, and 15 compute cores.
 # These could easily be changed in the function or added as arguments.
 # It does allow the user to specify the threshold of significant DMLs by providing the diff.methyl cut off and qvalue.
 # x : methylKit object made with the union function
 # dM :  difference in methylation among comparison.
 # q : q.value threshold.
 # cov : vector or matrix of covariates (this is used when all the data is examined, i.e. timepoint is included as covariate when looking
 # as temperature).
 # type : "DML" or "DMR" used to describe the state of the methylation datae (individual loci or regional)
 # comparison : "all","tp09","tp80" used to describe the specific comparison being performed
 # output :  output directory for the files.

performDiffMethSteps <- function(x,dM=50,q=0.01,cov=NULL,type="DML",comparison="all",output=baseOutputDir){
  newDir(output,paste0("dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2]))

  ### ChiSq Approach with overdispersion correction
  app <- "ChiSq_overDispCorrected"
  y<-calculateDiffMeth(x,
                       covariates=cov,
                       overdispersion="MN",test="Chisq",mc.cores=15)
  saveRDS(y,paste0(type,"_",comparison,app,".RData"))
  fwrite(y,paste0(type,"_",comparison,app,".csv"),sep = ",")
  
  # Multicomparison correction and critical % Diff. Threshold
  y_getDiff <- getMethylDiff(y,difference=50,qvalue=0.01)
  saveRDS(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.RData"))
  fwrite(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.csv"),sep = ",")
  # Hyper and Hypomethylation
  yhyper <- y_getDiff[y_getDiff$meth.diff > 0,]
  saveRDS(yhyper,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hyper.RData"))
  yhypo<- y_getDiff[y_getDiff$meth.diff < 0,]
  saveRDS(yhypo,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hypo.RData"))
  
  ### ChiSq Approach without overdispersion 
  app <- "ChiSq_uncorrected"
  y<-calculateDiffMeth(x,
                       covariates=cov,test="Chisq",mc.cores=15)
  saveRDS(y,paste0(type,"_",comparison,app,".RData"))
  fwrite(y,paste0(type,"_",comparison,app,".csv"),sep = ",")
  # Multicomparison correction and critical % Diff. Threshold
  y_getDiff <- getMethylDiff(y,difference=50,qvalue=0.01)
  saveRDS(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.RData"))
  fwrite(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.csv"),sep = ",")
  # Hyper and Hypomethylation
  yhyper <- y_getDiff[y_getDiff$meth.diff > 0,]
  saveRDS(yhyper,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hyper.RData"))
  yhypo<- y_getDiff[y_getDiff$meth.diff < 0,]
  saveRDS(yhypo,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hypo.RData"))
  
  ### DSS method
  # app <- "DSS"
  # y<-calculateDiffMethDSS(meth,mc.cores=15)
  # saveRDS(y,paste0(type,"_",comparison,app,".RData"))
  # fwrite(y,paste0(type,"_",comparison,app,".csv"),sep = ",")
  # # Multicomparison correction and critical % Diff. Threshold
  # y_getDiff <- getMethylDiff(y2,difference=50,qvalue=0.01)
  # saveRDS(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.RData"))
  # fwrite(y_getDiff,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_allSig.csv"),sep = ",")
  # # Hyper and Hypomethylation
  # yhyper <- y_getDiff[y_getDiff$meth.diff > 0,]
  # saveRDS(yhyper,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hyper.RData"))
  # yhypo<- y_getDiff[y_getDiff$meth.diff < 0,]
  # saveRDS(yhypo,paste0(type,"_",comparison,"_dM",dM,"_q",strsplit(as.character(q),"\\.")[[1]][2],app,"_hypo.RData"))
}

## Function for creates a simple bed file with chr,start,end,label from the cpgs matrix files generated from methylkit.
 # It defaults to putting this file in a folder called bed in a output directory specified by the user.
 # `x` : matrix with cpgs by row and labels columns including chr,start,end,and label (this can be predefined as anything in the matrix).
 # `start_offset` and `end_offset` can manually change the range of base offsets needed given the start and end base (this is to adjust
 # for the fact that methylkit and bismark defines the cpg as a single based even after destranding). The default behavior will 
 # adjust for this and create the appropriate start and end for a CpG.
 # `type` :  file name
 # `output` :  output directory
bedFileCreate <- function(x,start_offset=-1,end_offset=1,type = "test",output=NULL){
  y <- mutate(x,label=paste0(chr,"_",start+start_offset,"_",end+end_offset),
              start = start+start_offset, end = end+end_offset) %>% 
    select(chr, start, end,label) %>% 
    mutate_if(is.numeric, as.integer) #Save as a BED file, and avoid writing information in scientific notation
  fwrite(y,paste0(type,".bed"),
         sep = '\t',
         col.names = FALSE) #Save data as a BED file
}

## Summary plots for united methylKit data
 # Create a series of summary plots for evaluating methylation.
 # `x` = methylKit united object
 # `label` = Title name for each figure (default is no title)
 # `fileName` = File name (default is 'test')
 # `direc` = Directory for saving figures (default is current working directory)
summaryPlots <- function(x=NULL,labels="",fileName="test",direc=getwd()){
  # Reformat data from methylKit unite object for plotting
  xCov <- getData(x)[,seq(5,73,by=3)]
  xC <- getData(x)[,seq(6,73,by=3)]
  xBeta <- xC/xCov
  xBeta[is.na(xBeta)] <- 0
  colnames(xBeta) <- paste0(meta_final$ID,"_",meta_final$SFV)
  xBeta_summary <- data.frame(beta=rowMeans(xBeta),coverage=rowMeans(xCov))
  
  xBeta_meanList <- list()
  xCov_meanList <- list()
  j = 1
  label_df <- NULL
  xBeta_meanDF <- data.frame()
  xCov_meanDF <- data.frame()
  for(i in unique(meta_final$SFV)){
    xBeta_meanList[[j]] <- rowMeans(xBeta[,which(meta_final$SFV==i)])
    xCov_meanList[[j]] <- rowMeans(xCov[,which(meta_final$SFV==i)])
    label_df <- c(label_df,rep(i,times=length(xBeta_meanList[[j]])))
    j = j + 1
  }
  xBeta_meanDF <- ldply(xBeta_meanList, data.frame)
  colnames(xBeta_meanDF) <-  "beta"
  xBeta_meanDF$labels <- label_df
  xCov_meanDF <- ldply(xCov_meanList, data.frame)
  colnames(xCov_meanDF) <-  "coverage"
  xCov_meanDF$labels <- label_df

  ### Figures ###
  # Figure title
  title <- textGrob(paste0(labels,", Loci=",nrow(xBeta)),
                    gp=gpar(col="black", fontsize=20))
  ## Methylation Summary ##
  # General Methylation Histogram
  p1 <- ggplot(xBeta_summary,aes(x=beta*100)) + 
    geom_histogram(binwidth=1) + 
    geom_vline(xintercept = median(xBeta_summary$beta*100),colour="blue") +
    theme_cowplot() + 
    xlab("Mean Methylation")
  # Density Plot of Methylation by Treatment
  p2 <-  ggplot(xBeta_meanDF,aes(x=beta*100,colour=labels,group=labels)) + 
    geom_density(size=1.5) +
    theme_cowplot() + 
    scale_colour_manual(values = col_perm[c(1,3,2,4)],
                        labels=c("Control\n     D9","Control\n     D80","High OA\n     D9", "High OA\n     D80")) +
    xlab("Mean Methylation")
  mHist <- plot_grid(p1,p2,ncol=1,labels=c("A","B"))
  mHistF <- grid.arrange(arrangeGrob(mHist,
                                     top=title))
  ggsave(mHistF,filename = paste0(direc,"/",fileName,"_MethylationSummary.pdf"))
  ## Coverage Summary ##
  # General Histogram of Coverage
  p3 <- ggplot(xBeta_summary,aes(x=coverage)) + 
    geom_histogram(binwidth=1) + 
    geom_vline(xintercept = 5,colour="red") +
    geom_vline(xintercept = median(xBeta_summary$coverage),colour="blue") +
    theme_cowplot() + xlab("Mean Coverage")
  # Density Plot of Coverage by Treatment
  p4 <-  ggplot(xCov_meanDF,aes(x=coverage,colour=labels,group=labels)) + 
    geom_density(size=1.5) +
    theme_cowplot() + 
    scale_colour_manual(values = col_perm[c(1,3,2,4)],
                        labels=c("Control\n     D9",
                                 "Control\n     D80",
                                 "High OA\n     D9",
                                 "High OA\n     D80")) +
    xlab("Mean Coverage")
  cHist <- plot_grid(p3,p4,ncol=1,labels=c("A","B"))
  cHistF <- grid.arrange(arrangeGrob(cHist,
                                     top=title))
  ggsave(cHistF,filename = paste0(direc,"/",fileName,"_CoverageSummary.pdf"))
  ## Summary Methylation vs Coverage Comparison ##
  # Comparison Methylation by Coverage (plot loci density)
  p5 <- ggplot(xBeta_summary,aes(x=beta*100,y=coverage)) + 
    geom_bin2d(bins=100) +
    theme_cowplot() +
    labs(x = "Mean Methylation",y="Mean Coverage")
  # Comparison Methylation by Coverage (binned medians)
  p6 <- ggplot(xBeta_summary,aes(x=beta*100,y=coverage)) + 
    stat_summary_bin(fun.y = "median", geom="point",
                     binwidth = 5,size=5) +
    theme_cowplot() +
    labs(x = "Mean Methylation",y="Mean Coverage (median for 5% increments)")
  comp <- plot_grid(p5,p6,nrow = 2,ncol=1,labels = c("A","B"))
  compF <- grid.arrange(arrangeGrob(comp,
                                    top=title))
  ggsave(compF,filename = paste0(direc,"/",fileName,"_ComparisonSummary.pdf"))
  ## Create PCA with all individuals
  pca <- prcomp(t(xBeta))
  eigs <- pca$sdev^2
  temp_df <- data.frame(x=pca$x[,1],y=pca$x[,2],condition=meta_final$SFV)

  p_temp <- ggplot(temp_df,aes(x=x,y=y,colour=condition,shape=condition)) +
    geom_point(size=4) + theme_cowplot() +
    scale_colour_manual(values = col_perm[c(1,3,2,4)],
                        labels=c("Control D9","Control D80","High OA D9","High OA D80")) +
    scale_shape_manual(values = c(15,15,17,17),
                       labels=c("Control D9","Control D80","High OA D9","High OA D80")) +
    labs(title=paste0(labels,", Loci=",nrow(xBeta)),colour="",shape="") +
    theme(plot.title = element_text(hjust = 0.5))
  ggsave(p_temp,filename=paste0(direc,"/",fileName,"_PCA.pdf"))
  
  ## Correlation plots among individuals ##
  pdf(file = paste0(direc,"/",fileName,"_CorrelationCirclesSummary.pdf"))
  corrplot(cor(xBeta), diag = FALSE, order = "hclust",
            tl.pos = "td",  tl.cex = 0.8,tl.col="black",
            method = "circle", type = "upper",cl.cex = 0.9,
            col = brewer.pal(n = 8, name = "PRGn"))
  dev.off()
  pdf(paste0(direc,"/",fileName,"_CorrelationNumericSummary.pdf"))
  corrplot(cor(xBeta), diag = FALSE, order = "hclust",
            tl.cex = (0.8), tl.srt = 45,tl.col="black",
            method = "number",
            type = "lower",cl.cex = (0.9),number.cex = (0.6),
            col = brewer.pal(n = 8, name = "PRGn"))
  dev.off()
}

## Methylation plots by treatment
# `x` : methylObject generated by methylKit unite.
# `fileName` : Name on the file to represent the comparison for the specific set of DMLs
# `label` :  dataframe that contains a column of sample IDs and column of treatments.
methylationPlotByTreatment <- function(x,fileName,label){
  temp_t = matrix(unlist(getData(x)[,seq(5,ncol(x),by=3)]),ncol=nrow(label))
  temp_c = matrix(unlist(getData(x)[,seq(6,ncol(x),by=3)]),ncol=nrow(label))
  temp_b = temp_c/temp_t
  print("Making figures...")
  for(i in 1:nrow(x)){
    plot_df <- data.frame(ID=label$ID,meta=label$var,b=temp_b[i,],cov=temp_t[i,]) 
    print(paste0("Plot ",i," of ",nrow(x)))
    # Basic Boxplot by treatment
    p1 <- ggplot(plot_df,aes(x=meta,y=b,group=meta,col=meta))+ geom_boxplot() + labs(y="Prop. Methylation")
    # Methylation for each individual
    p2 <- ggplot(plot_df,aes(x=ID,y=b,group=meta,col=meta))+ geom_point() + labs(y="Prop. Methylation",x="Individuals") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1))
    # Coverage plot
    p3 <- ggplot(plot_df,aes(x=cov,y=b,col=meta))+ geom_point() + 
      labs(y="Prop. Methylation",x="Coverage")
    # Coverage histogram
    p4 <- ggplot(plot_df,aes(x=cov,fill=meta)) + geom_histogram(bins=30)
    final<-plot_grid(p1,p2,p3,p4)#,nrow=2 # p1,p2,
    save_plot(paste0("DMLbyTrt_",fileName,"_",getData(x)[i,1],"_",getData(x)[i,2],"_beta.png"),final, ncol = 2)
  }
}

#### Data ####

## Set working directory
# Paths to input and output directories (from Github repo)
inputDir_meta <- "~/Github/AE17_Cvirginica_MolecularResponse"
inputDir_CytoSummaries <- "/path/to/CytoReportFiles"
# Outputs on local directory
outputDir <- "/home/downeyam/Github/AE17_Cvirginica_MolecularResponse/data/MBDBS_seq"
setwd(inputDir_meta)
meta <- readRDS("data/meta/metadata_20190811.RData")

# Pathes on computing cluster
inputDir_meta <- "/shared_lab/20180226_RNAseq_2017OAExp/DNAm/"
inputDir_CytoSummaries <- "/shared_lab/20180226_RNAseq_2017OAExp/DNAm/processed_samples/03_CytoSummaries"
setwd(inputDir_meta)
meta <- readRDS("metadata/metadata_20190811.RData")
## Data Manipulation and READ IN
# Meta data
meta_final <- meta[meta$ID != "17099",] # This individual was removed due to poor sequencing
# Convert treatment into binary (0,1)
trt <- ifelse(meta_final$treatment==400,0,1)
timeVec <- ifelse(meta_final$Time=="09",0,1)

# Cytosine Summary Reports from Bismark
setwd(inputDir_CytoSummaries)
file.list <- list.files(pattern="17")

# Removing sample 17099 which sequenced poorly
file.list <- file.list[-11]
ID <- meta_final$ID

#### Read in files ####

## Stores all CpGs (mincov -  minimum coverage = 0)
myobj <- methRead(as.list(file.list),
                 sample.id=as.list(ID),
                 assembly="t2",
                 treatment=trt, # Preliminary calling of the treatment
                 mincov = 0, # minimum coverage = 0 (filtering happens separately)
                 pipeline='bismarkCytosineReport') # I use outputs directly from bismark
# Here I want to be able to see all CpGs, and we also filter in the next step

# Here I change my working directory since we will be saving our outputs from here on out.
newDir(inputDir_CytoSummaries,"methyKitOutput")
baseOutputDir <- getwd()

#Saving raw in methReadObj
saveRDS(myobj,"methylKitObj_unfiltered.RData")
fwrite(myobj,"methylKitObj_unfiltered.csv",sep = ",")

#### Filter samples ####
# Defaults used
# Needs at least 5x coverage
# And remove loci with coverage > 200 (potential PCR bias)
filtered.myobj <- filterByCoverage(myobj,lo.count=5,lo.perc=NULL,
                                hi.count=200,hi.perc=NULL)

#### Normalize counts ####
 # Thoughts on normalization: methylKit comes with a simple normalizeCoverage() function
 # to normalize read coverage distributions between samples. Ideally, you should first 
 # filter bases with extreme coverage to account for PCR bias using filterByCoverage() 
 # function, then run normalizeCoverage() function to normalize coverage between samples.
 # These two functions will help reduce the bias in the statistical tests that might 
 # occur due to systematic over-sampling of reads in certain samples.

 # The function normalizes coverage values between samples using a scaling factor derived 
 # from differences between mean or median of coverage distributions
 # Default normalization is median
## Normalizing filter (5X) object
norm.filtered.myobj <- normalizeCoverage(filtered.myobj,method="median") 
saveRDS(norm.filtered.myobj,"methylKitObj_cov5Filtered_medianNormalized.RData")
fwrite(norm.filtered.myobj,"methylKitObj_cov5Filtered_medianNormalized.csv",sep = ",")
## Normalizing unfiltered object 
norm.myobj <- normalizeCoverage(myobj,method="median")
saveRDS(norm.myobj,"methylKitObj_unfiltered_medianNormalized.RData")
fwrite(norm.myobj,"methylKitObj_unfiltered_medianNormalized.csv",sep = ",")

#### Unite will combine samples ####
 # Function will shrink dataframe of each individual to only 
 # those shared by all individuals and will be used create a matrix
 # sample x loci matrix

 # Since we are focusing on CpGs we used the destrand = TRUE arguement to combine coverage from cytosines
 # within a CpG from eith strand. This increases the number of loci in our final data, but makes the important
 # assuming that both strands have the same methylation state (methylated or unmethylated).

# All loci (unfiltered) and no normalization
meth_all <- unite(myobj, destrand=TRUE)
saveRDS(meth_all,"methylKitObj_unfiltered_united.RData")
summaryPlots(x=meth_all,label = "Unfiltered and Not Normalized",
             fileName = "unfiltered_nonNormalized")
fwrite(meth_all,"methylKitObj_unfiltered_united.csv",sep = ",")
## Unfiltered and normalized object
meth_norm_all <- unite(norm.myobj, destrand=TRUE)
saveRDS(meth_norm_all,"methylKitObj_unfiltered_medianNormalized_united.RData")
summaryPlots(x=meth_norm_all,label = "Unfiltered and Median Normalized",
             fileName = "unfiltered_medianNormalized")
fwrite(meth_norm_all,"methylKitObj_unfiltered_medianNormalized_united.csv",sep = ",")
## Loci with at least 5X coverage (filtered) for ALL individuals and normalized
meth_filt <- unite(norm.filtered.myobj, destrand=TRUE)
saveRDS(meth_filt,"methylKitObj_cov5Filtered_medianNormalized_united.RData")
summaryPlots(x=meth_filt,label = "x5 Filter Threshold and Median Normalized",
             fileName = "cov5Filtered_medianNormalized")
fwrite(meth_filt,"methylKitObj_cov5Filtered_medianNormalized_united.csv",sep = ",")

## Alternative filtering ##
 # Here I take a custom filtering approach similar to what was done with GE. I
 # retain all CpGs that have 5x coverage for at least 5 individuals (4 for control day 9)
 # in at least one treatment x time combination (i.e. Day 80 Treatment 2800). The rationale
 # here is we could have methylation occuring in only one treatment X time combination, which
 # might be of particular interest, since it represents differential response. However, the
 # traditional filtering would remove those loci. This is problematic since our 
 # sequencing method (MBD-BSseq) is biased toward representing methylated regions. This means
 # that lack of coverage could be the result of no methylation.

# meth_all <- readRDS("methylKitObj_unfiltered_united.RData")
meth_get <- getData(meth_all) # Acquire sampleXloci matrix from methylKitObj
meth_cov <- meth_get[,seq(from=5,to=73,by=3)] # Select columns for coverage
meth_cyto <- meth_get[,seq(from=6,to=73,by=3)] # select columns for Cytosines
meth_beta <- meth_cyto/meth_cov # Calculate beta (% methylation) as the number of cytosines/total coverage
meth_beta(is.na(meth_beta)) <- 0 # Coverts any NaNs resulting from dvidings by 0
# Breaking down expression coverage by treatment*time combination
#Day 9 Trt 2800
keep_D9.2800 <- rowSums(meth_cov[,meta_final$SFV=="09.2800"] > 5) >= 5
#Day 9 Trt 400
keep_D9.400 <- rowSums(meth_cov[,meta_final$SFV=="09.400"] > 5) >= 5
#Day 80 Trt 2800
keep_D80.2800 <- rowSums(meth_cov[,meta_final$SFV=="80.2800"] > 5) >= 5
#Day 80 Trt 400
keep_D80.400 <- rowSums(meth_cov[,meta_final$SFV=="80.400"] > 5) >= 4
keep_gene <- rowSums(cbind(keep_D9.2800,keep_D9.400,
                           keep_D80.2800,keep_D80.400)) >= 1
meth_altCov <- meth_all[keep_gene, ]
saveRDS(meth_altCov,"methylKitObj_5x5SampleSingleTreatment_united.RData")
summaryPlots(x=meth_altCov,label = "x5 Filter 5 Samples Single Trt and non Normalized",
             fileName = "alt5xFilter_nonNormalized")

#### Use reorganize to subset data by timepoints ####
 # Subsetting data so we can do targetted analysis we samples at each day or within 
 # a particular treatment.
outputDir <- "/shared_lab/20180226_RNAseq_2017OAExp/DNAm/processed_samples/03_CytoSummaries/methyKitOutput"
setwd(outputDir)
meth_filt <- readRDS("methylKitObj_cov5Filtered_medianNormalized_united.RData")

## All time points with time as variable 
meth_time <- meth_filt
meth_time@treatment <-  timeVec
meta_time

 # NOTE: This was done using filtered (5x) and median normalized data
## Day 9 (5x coverage)
meta_9 <- meta_final[meta_final$Time == "09",]
trt9 <-  ifelse(meta_9$treatment==400,0,1)
meth9 <- reorganize(meth_filt,sample.ids=meta_9$ID,treatment=trt9)
saveRDS(meth9,"methylKitObj_cov5Filtered_medianNormalized_united_TP9.RData")
fwrite(meth9,"methylKitObj_cov5Filtered_medianNormalized_united_TP9.csv",sep = ",")
## Day 80 (5x coverage)
meta_80 <- meta_final[meta_final$Time == "80",]
trt80 <- ifelse(meta_80$treatment==400,0,1)
meth80 <- reorganize(meth_filt,sample.ids=meta_80$ID,treatment=trt80)
saveRDS(meth80,"methylKitObj_cov5Filtered_medianNormalized_united_TP80.RData")
fwrite(meth80,"methylKitObj_cov5Filtered_medianNormalized_united_TP80.csv",sep = ",")
## Control Treatment - Day 9 vs Day 80 (5x coverage)
meta_C <- meta_final[meta_final$treatment == 400,]
trtC <- ifelse(meta_C$Time=="09",0,1)
methC <- reorganize(meth_time,sample.ids=meta_C$ID,treatment=trtC)
saveRDS(methC,"methylKitObj_cov5Filtered_medianNormalized_united_control.RData")
fwrite(methC,"methylKitObj_cov5Filtered_medianNormalized_united_control.csv",sep = ",")
## Exposed Treatment - Day 9 vs Day 80 (5x coverage)
meta_E <- meta_final[meta_final$treatment == 2800,]
trtE <- ifelse(meta_E$Time=="09",0,1)
methE <- reorganize(meth_time,sample.ids=meta_E$ID,treatment=trtE)
saveRDS(methE,"methylKitObj_cov5Filtered__medianNormalized_united_exposed.RData")
fwrite(methE,"methylKitObj_cov5Filtered__medianNormalized_united_exposed.csv",sep = ",")

#### Create separate counts matrices and beta (methylC/totalC) ####
#newDir(baseOutputDir,"count_matrix_5X")

meth_all_counts <- as.matrix(getData(meth_filt)[,5:73])
#Generate the row names 
rName <- paste0(getData(meth_filt)[,1],"_",getData(meth_filt)[,2])
# Generate the counts matrix
meth_total <-meth_all_counts[,seq(from=1,to=69,by=3)]
meth_M <-meth_all_counts[,seq(from=2,to=69,by=3)]
meth_uM <-meth_all_counts[,seq(from=3,to=69,by=3)]
meth_beta <- meth_M/meth_total

rownames(meth_total) <- rName
rownames(meth_M) <- rName
rownames(meth_uM) <- rName
rownames(meth_beta) <- rName

fwrite(meth_total,"countMatrix_cov5Filtered_medianNormalized_totalCounts.csv",sep = ",")
fwrite(meth_M,"countMatrix_cov5Filtered_medianNormalized_methylCCounts.csv",sep = ",")
fwrite(meth_uM,"countMatrix_cov5Filtered_medianNormalized_unMethylCCounts.csv",sep = ",")
fwrite(meth_beta,"countMatrix_cov5Filtered_medianNormalized_beta.csv",sep = ",")

#### Create a summary of methylation for each CpG ####
# Use the beta matrix created in previous section
b <- as.matrix(meth_beta)
# Mean methylation all samples
mean_beta <- rowMeans(b)
sd_beta <- rowSds(b)
cv_beta <- sd_beta/mean_beta

# Individual Treatment x Time level combination summaries
# Mean methylation Control day 9
mean_9C <- rowMeans(b[,meta_final$SFV == "09.400"]) # mean
sd_9C <- rowSds(b[,meta_final$SFV == "09.400"]) # standard deviation
cv_9C <- sd_9C/mean_9C # coefficient of variation
# Mean methylation Exposed day 9
mean_9E <- rowMeans(b[,meta_final$SFV == "09.2800"])
sd_9E <- rowSds(b[,meta_final$SFV == "09.2800"])
cv_9E <- sd_9E/mean_9E
# Mean methylatoin Control day 80
mean_80C <- rowMeans(b[,meta_final$SFV == "80.400"])
sd_80C <- rowSds(b[,meta_final$SFV == "80.400"])
cv_80C <- sd_80C/mean_80C
# Mean methylation Exposed day 80
mean_80E <- rowMeans(b[,meta_final$SFV == "80.2800"])
sd_80E <- rowSds(b[,meta_final$SFV == "80.400"])
cv_80E <- sd_80E/mean_80E

# Coefficient of Variation among treatments (all levels)
mean_AmongTrt <- cbind(mean_9C,mean_9E,mean_80C,mean_80E)
sd_mean_AmongTrt <- rowSds(mean_AmongTrt)
mean_mean_AmongTrt <- rowMeans(mean_AmongTrt)
cv_mean_AmongTrt <- sd_mean_AmongTrt/mean_mean_AmongTrt

# Difference in Methylation trt
mean_C <- rowMeans(b[,meta_final$Treatment ==400])
mean_E <- rowMeans(b[,meta_final$Treatment == 2800])
diff_Trt <- mean_E-mean_C

# Difference in Methylation time
mean_9 <- rowMeans(b[,meta_final$Time == "09"])
mean_80 <- rowMeans(b[,meta_final$Time == "80"])
diff_Time <- mean_E-mean_C

# Difference in Methylation tp9
diff_tp9_Trt <- mean_9E-mean_9C
# Difference in Methylation tp80
diff_tp80_Trt <- mean_80E-mean_80C
# Creat matrix of all CpG summary stats
meth_summary <- cbind(mean_beta,
                      mean_9C,mean_9E,mean_80C,mean_80E,
                      sd_9C,sd_9E,sd_80C,sd_80E,sd_mean_AmongTrt,
                      cv_9C,cv_9E,cv_80C,cv_80E,cv_mean_AmongTrt,
                      diff_Trt,diff_Time,diff_tp9_Trt,diff_tp80_Trt)
meth_summary[is.na(meth_summary)]<-0 
# in cases where the mean of treatmentxtime combination is 0 for the CV calculation will produce an NA,
# we want to change these to 0s.
    saveRDS(meth_summary,"methylSummary_cov5Filtered_medianNormalized_united.RData")

#### Create 'bed' file formats for these outputs ####
newDir(baseOutputDir,"bed")
meth_all <- readRDS("methylKitObj_unfiltered_united.RData")
meth_filt <- readRDS("methylKitObj_cov5Filtered_medianNormalized_united.RData")
meth_altCov <- readRDS("methylKitObj_5x5SampleSingleTreatment_united.RData")

## All CpGs 
bedFileCreate(x=meth_all,type="CpG_noFilter_noNormalization",output=baseOutputDir)
## CpGs with 5x coverage with summary 
bedFileCreate(x=meth_filt,type="CpG_5xCov_medianNormalization",output=baseOutputDir)
## CpGs after alternative filtration and no normalization
bedFileCreate(x=meth_altCov,type="CpG_5x5SampleSingleTreatment_noNormalization",output=baseOutputDir)

# meth <- readRDS(paste0(baseOutputDir,"/methylKitObj_cov5Filtered_medianNormalized_united.RData"))
# CpG_5xCov_bed <- mutate(meth,label=paste0(chr,"_",start),start = start-1, end = end + 1) %>% 
#   select(chr, start, end,label) %>% 
#   mutate_if(is.numeric, as.integer) #Save as a BED file, and avoid writing information in scientific notation
# CpG_5xCov_bed <- data.frame(CpG_5xCov_bed,meth_summary)
# fwrite(CpG_5xCov_bed,
#        "CpG_5xCov.bed",
#        sep = '\t',
#        col.names = FALSE) #Save data as a BED file

#### Read in files from saved folder (optional)####
# If you want to start with the united methyKit Objects 
 setwd(outputDir)
 baseOutputDir <- getwd()
 meth_all <- readRDS("methylKitObj_unfiltered_medianNormalized_united.RData")
 meth <- readRDS("methylKitObj_cov5Filtered_medianNormalized_united.RData")
 meth9 <- readRDS("methylKitObjcov5Filtered_medianNormalized_united_TP9.RData")
 meth80 <- readRDS("methylKitObj_cov5Filtered_medianNormalized_united_TP80.RData")

#### Differential Methyl ####
# Differential Methylation done with the calculateDiffMeth() function, which takes a logisitic regression approach
# Critical criteria for significant
  # q value : 0.01
  # Diff methylation : >= 50%

cov = data.frame(tp=meta_final$timepoint)
# All trt
performDiffMethSteps(meth,cov=cov,type="DML",output=baseOutputDir)
# All tp 9
performDiffMethSteps(meth9,type="DML",comparison="tp9",output=baseOutputDir)
# All tp80
performDiffMethSteps(meth80,type="DML",comparison="tp80",output=baseOutputDir)
# All timepoint
cov = data.frame(tp=meta_final$treatment)
performDiffMethSteps(meth_time,cov=cov,type="DML",comparison="all_timepoint",output=baseOutputDir)
# C
performDiffMethSteps(methC,type="DML",comparison="C",output=baseOutputDir)
# E
performDiffMethSteps(methE,type="DML",comparison="E",output=baseOutputDir)

## DMLs all timepoints
files <- list.files(paste0(baseOutputDir,"/dM50_q01"))
subsetSample <- c("all_d","tp9","tp09","tp80","all_t","C","E")
for(i in 1:length(subsetSample)){
  target <- files[grep(subsetSample[i],files)]
  target <- target[grep("DML",target)]
  temp <- target[grep(".RData",target)]
  if(length(temp)>0){
    for(j in 1:length(temp)){
      x <- readRDS(paste0(baseOutputDir,"/dM50_q01/",temp[j]))
      name <- strsplit(temp[j],"\\.")[[1]][1]
      bedFileCreate(x=x,type = name,output=baseOutputDir)
    }
  }
  target <- files[grep(subsetSample[i],files)]
  target <- target[grep("DMR",target)]
  temp <- target[grep(".RData",target)]
  if(length(temp)>0){
    for(j in 1:length(temp)){
      x <- readRDS(paste0(baseOutputDir,"/dM50_q01/",temp[j]))
      name <- strsplit(temp[j],"\\.")[[1]][1]
      bedFileCreate(x=x,type = name,output=baseOutputDir)
    }
  }
}





#### DMRs - Tiling (optional) #####
# This will section will perform a similar differential methylation analysis, 
# but summarizing CpGs into 100bp windows.
tiles <- tileMethylCounts(norm.myobj,win.size=100,step.size=100)
saveRDS(tiles,"methylKitObj_unfiltered_medianNormalized_tiled.RData")

## Unite samples
# All points
meth_all_tile <- unite(tiles, destrand=TRUE)
saveRDS(meth_all_tile,"methylKitObj_unfiltered_medianNormalized_tiled_united.RData")

# Read in data (optional)
# newDir(inputDir_CytoSummaries,"methylObjFiles")
# baseOutputDir <- getwd()
# meth_all_tile <- readRDS("methylKitObj_all_medianNormalizedUnfiltered_united_tiled.RData")

# All points with time rather than OA as treatment
meth_all_time_tile <- meth_all_tile
meth_all_time_tile@treatment <- timeVec 
# TP 9
meta_9 <- meta_final[meta_final$Time == "09",]
trt9 <-  ifelse(meta_9$treatment== 400,0,1)
meth9_tile <- reorganize(meth_all_tile,sample.ids=meta_9$ID,treatment=trt9)
# TP 80
meta_80 <- meta_final[meta_final$Time == "80",]
trt80 <- ifelse(meta_80$treatment==400,0,1)
meth80_tile <- reorganize(meth_all_tile,sample.ids=meta_80$ID,treatment=trt80)
# Control Treatment (day 9  vs day 80)
meta_C <- meta_final[meta_final$treatment == 400,]
trtC <- ifelse(meta_C$Time=="09",0,1)
methC_tile <- reorganize(meth_all_time_tile,sample.ids=meta_C$ID,treatment=trtC)
# Exposure Treatment (day 9  vs day 80)
meta_E <- meta_final[meta_final$treatment == 2800,]
trtE <- ifelse(meta_E$Time=="09",0,1)
methE_tile <- reorganize(meth_all_time_tile,sample.ids=meta_E$ID,treatment=trtE)

#### Differential Methylation Tiles (optional) ####
# Get all loci that are differentially methylated
# with a difference of at least 50% and a qvluae=0.01

cov = data.frame(tp=meta_final$timepoint)
# All trt
performDiffMethSteps(meth_all_tile,cov=cov,type="DMR",output=baseOutputDir)
# All tp 9
performDiffMethSteps(meth9_tile,type="DMR",comparison="tp09",output=baseOutputDir)
# All tp80
performDiffMethSteps(meth80_tile,type="DMR",comparison="tp80",output=baseOutputDir)
# All timepoint
cov = data.frame(tp=meta_final$treatment)
performDiffMethSteps(meth_all_time_tile,cov=cov,type="DMR",comparison="all_timepoint",output=baseOutputDir)
# C
performDiffMethSteps(methC_tile,type="DMR",comparison="C",output=baseOutputDir)
# E
performDiffMethSteps(methE_tile,type="DMR",comparison="E",output=baseOutputDir)

## All Regions 
meth_all_tile <- readRDS(paste0(baseOutputDir,"/methylKitObj_unfiltered_medianNormalized_tiled_united.RData"))
bedFileCreate(x=meth_all_tile,type="Region_100BpTile",output=baseOutputDir)
